pkg/reconciler/build/timeouts_handler.go

```diff
@@ -0,0 +1,81 @@
+package build
+
+import (
+	"time"
+
+	clientset "github.com/knative/build/pkg/client/clientset/versioned"
+	"go.uber.org/zap"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/kubernetes"
+)
+
+const checkInterval = 10 * time.Second
+
+// TimeoutSet contains required objects to handle build timeouts
+type TimeoutSet struct {
+	logger         *zap.SugaredLogger
+	kubeclientset  kubernetes.Interface
+	buildclientset clientset.Interface
+	stopCh         <-chan struct{}
+}
+
+// NewTimeoutHandler returns TimeoutSet filled structure
+func NewTimeoutHandler(logger *zap.SugaredLogger,
+	kubeclientset kubernetes.Interface,
+	buildclientset clientset.Interface,
+	stopCh <-chan struct{}) *TimeoutSet {
+	return &TimeoutSet{
+		logger:         logger,
+		kubeclientset:  kubeclientset,
+		buildclientset: buildclientset,
+		stopCh:         stopCh,
+	}
+}
+
+// TimeoutHandler walks through all namespaces, gets list of builds and checks timeouts
+func (t *TimeoutSet) TimeoutHandler() error {
+	ticker := time.NewTicker(checkInterval)
+	defer ticker.Stop()
+
+	for {
+		select {
+		case <-t.stopCh:
+			return nil
+		case <-ticker.C:
+			namespaces, err := t.kubeclientset.CoreV1().Namespaces().List(metav1.ListOptions{})
+			if err != nil {
+				t.logger.Errorf("%s", err)
+				continue
+			}
+			for _, namespace := range namespaces.Items {
+				if err := t.checkPodTimeouts(namespace.GetName()); err != nil {
+					t.logger.Errorf("%s", err)
+				}
+			}
+		}
+	}
+}
+
+func (t *TimeoutSet) checkPodTimeouts(namespace string) error {
+	builds, err := t.buildclientset.BuildV1alpha1().Builds(namespace).List(metav1.ListOptions{})
+	if err != nil {
+		return err
+	}
+	for _, build := range builds.Items {
+		if isDone(&build.Status) {
+			continue
+		}
+		r := Reconciler{
```

Comment:
I don't really like having two `Reconciler` instances involved at the same time, I think we'll get confused about which one is responsible for what.

How about we expose `TimeoutSet.CheckTimeout(build)` and inside the `Reconcile` method (which is called on every pod event) we check whether the build has timed out since the last event.

Separate from that, a `TimeoutSet` is also running goroutines which wait in the background for the timeout to expire and update builds, but just in case those don't happen we'll at least be able to enforce the timeout on the next pod event.